File: Fibonacci/Fibonnaci.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

public class Fibonnaci{

    public static List<Integer> mutable(int position)
    {   
        List<Integer> list = new ArrayList<Integer>();
        for(int currentPosition = 0 ; currentPosition < position ; currentPosition++)
        {  
            if(currentPosition <= 1)
              list.add(1);
            else
              list.add(list.get(currentPosition - 1) + list.get(currentPosition - 2));
        }
        return list;
    }

    public static void recursive(int position , List<Integer> fibRecMutable)
    {   
        List<Integer> fibTemp = fibRecMutable;
        int size = fibTemp.size();
        if(size < position)
        {
           if(size <= 1)
              fibTemp.add(1);
           else 
              fibTemp.add(fibTemp.get(size - 1) + fibTemp.get(size - 2));
           recursive(position , fibTemp);
        }
    }

    public static List<Integer> immutable(int position)
    {    
        List<Integer> fibImmutable;
        if(position == 1)
           fibImmutable = Stream.of(1).collect(Collectors.toList());
        else
        {
           fibImmutable = Stream.iterate(2 , i -> i + 1)
                                .limit(position - 2)
                                .reduce(Stream.of(1 , 1) , (a , b) -> process(a , b) , Stream::concat)
                                .collect(Collectors.toList());
        }
        
        return (fibImmutable);
    }
     
    public static Stream process(final Stream<Integer> tempStream, final int index)
    {
       final List<Integer> tempList = tempStream.collect(Collectors.toList());
       final Stream<Integer> nextStream = Stream.concat(tempList.stream() , Stream.of(tempList.stream().skip(index-2).reduce(0,Integer::sum)));
       return nextStream;
    }

    public static void main(String[] args){
        List<Integer> fibTemp = new ArrayList<Integer>();
        int position = 7;
        if(position > 0)
        {
          System.out.println("Fibanocci Recursive ");
          recursive(position , fibTemp);
          for (int value : fibTemp)
            System.out.print(value + " ");
          System.out.println();
          fibTemp.clear();
          System.out.println("Fibanocci Mutable");
          fibTemp = mutable(position);
          for(int value : fibTemp)
            System.out.print(value + " ");
          System.out.println();      
          fibTemp.clear();
          fibTemp = immutable(position);
          System.out.println("Fibanocci Immutable");
          for (int value : fibTemp)
            System.out.print(value + " ");
          System.out.println();
        }
        else
            System.out.println("Oops! Try again with a valid position");
    }
}

File: Fibonacci/fibonnaci.rb
def fibonnaci_recursion(position , fibSeriesRec)
 tempList = fibSeriesRec
 currentPosition = fibSeriesRec.size
 if currentPosition < position
  if currentPosition <= 1
    tempList << 1
  else
    tempList << (tempList[currentPosition - 1] + tempList[currentPosition - 2])
  end
  fibonnaci_recursion(position , tempList)
 end
end

def fibonnaci_mutable(position , fibMutable)
  fibSeriesMutable = fibMutable
  for currentPosition in 0..position-1
    if currentPosition <= 1
        fibMutable << 1;
    else
        fibMutable << (fibMutable[currentPosition - 1] + fibMutable[currentPosition - 2])
    end
  end
end

def fibonnaci_immutable(position)
  if position == 1
   (1).inject(0){|result , index| index}
  else
   (2..position - 1).inject([1 , 1]){|result , index| result<<(result[index - 2] + result[index - 1])}
  end
end

def main()
fibTemp = []
position = 7
 if  position > 0
  puts "\nFibonnaci series of position #{position} :"
  #By Recursion
  puts "---> Implemented by using Recurssion is "
  fibonnaci_recursion(position , fibTemp)
  fibTemp.each {|value|  print value.to_s+" "}
  fibTemp.clear
  puts
  #By not using recurssion
  puts "---> Implemented with out using Recurssion is "
  fibonnaci_mutable(position , fibTemp)
  fibTemp.each {|value|  print value.to_s+" "}
  fibTemp.clear
  puts
  #Immutable
  puts "---> Implemented employing immutability is "
  fibTemp = fibonnaci_immutable(position)
  fibTemp.each {|value|  print value.to_s+" "}
  fibTemp.clear
  puts
else
  puts "Oops! Try again with a valid length"
end
end

main()
